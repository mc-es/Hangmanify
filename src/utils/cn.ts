import type { ClassValue } from 'clsx';
import clsx from 'clsx';
import { twMerge } from 'tailwind-merge';

/**
 * ### Combines conditional Tailwind class names and resolves conflicts
 *
 * @param {...ClassValue[]} inputs - Class names, booleans, objects, or arrays
 * @returns {string} Merged and deduplicated class string
 * @example
 * cn('p-4', isDark && 'bg-black') // 'p-4 bg-black' if isDark is true, otherwise 'p-4'
 */
export const cn = (...inputs: ClassValue[]): string => twMerge(clsx(...inputs));
