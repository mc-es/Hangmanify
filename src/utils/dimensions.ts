import { Dimensions as NativeDimensions, PixelRatio } from 'react-native';

const { width: deviceWidth, height: deviceHeight } = NativeDimensions.get('window');

const guidelineBaseWidth = 375;
const guidelineBaseHeight = 812;
const BLEND_ZONE = 5;

const BREAKPOINTS = {
  '2xs': { size: 320, scale: 0.65 },
  'xs-2xs': { size: 340, scale: 0.7 },
  xs: { size: 360, scale: 0.75 },
  'xs-sm': { size: 375, scale: 0.75 },
  sm: { size: 390, scale: 0.85 },
  'sm-md': { size: 420, scale: 0.9 },
  md: { size: 480, scale: 1 },
  'md-lg': { size: 540, scale: 1.05 },
  lg: { size: 600, scale: 1.1 },
  'lg-xl': { size: 680, scale: 1.15 },
  xl: { size: 760, scale: 1.2 },
  'xl-2xl': { size: 800, scale: 1.3 },
  '2xl': { size: 840, scale: 1.4 },
} as const;

const easeIn = (t: number): number => t * t;
const easeOut = (t: number): number => t * (2 - t);

const getScale = (width: number): number => {
  const breakpoints = Object.values(BREAKPOINTS).sort((x, y) => x.size - y.size);

  if (width <= breakpoints[0].size) return breakpoints[0].scale;
  if (width >= breakpoints[breakpoints.length - 1].size)
    return breakpoints[breakpoints.length - 1].scale;

  for (let i = 0; i < breakpoints.length - 1; i++) {
    const current = breakpoints[i];
    const next = breakpoints[i + 1];

    if (width >= current.size && width < next.size) {
      const scaleRange = next.scale - current.scale;
      const widthRange = next.size - current.size;
      const delta = width - current.size;

      let finalProgress;

      if (delta < BLEND_ZONE) {
        const t = delta / BLEND_ZONE;
        finalProgress = easeIn(t) * (BLEND_ZONE / widthRange);
      } else if (delta > widthRange - BLEND_ZONE) {
        const t = (delta - (widthRange - BLEND_ZONE)) / BLEND_ZONE;
        finalProgress =
          (widthRange - BLEND_ZONE) / widthRange + easeOut(t) * (BLEND_ZONE / widthRange);
      } else finalProgress = delta / widthRange;

      return current.scale + scaleRange * finalProgress;
    }
  }

  return 1;
};

const scale = getScale(deviceWidth);

// horizontal & vertical scale
const hs = (size: number): number => (deviceWidth / guidelineBaseWidth) * size * scale;
const vs = (size: number): number => (deviceHeight / guidelineBaseHeight) * size;

// width & height percent
const wp = (percent: number): number => (deviceWidth * percent) / 100;
const hp = (percent: number): number => (deviceHeight * percent) / 100;

// moderate scale
const ms = (size: number, factor = 0.5): number => size + (hs(size) - size) * factor;

// font scale
const fs = (size: number): number => {
  const fontScale = PixelRatio.getFontScale();
  return Math.round(PixelRatio.roundToNearestPixel(hs(size) * fontScale));
};

export const Dimensions = { hs, vs, wp, hp, ms, fs };
