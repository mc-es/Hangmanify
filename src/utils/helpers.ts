import type { TranslateOptions } from 'i18n-js';

function getNestedValue<T>(obj: T, key: string): unknown {
  return key.split('.').reduce((acc: unknown, curr: string): unknown => {
    if (acc !== null && typeof acc === 'object' && curr in acc)
      return (acc as Record<string, unknown>)[curr];

    return null;
  }, obj as unknown);
}
function interpolate(template: string, options: TranslateOptions): string {
  return template.replace(/{{\s*(\w+)\s*}}/g, (_match, key) =>
    key in options ? String(options[key]) : ''
  );
}

export const helpers = { getNestedValue, interpolate };
