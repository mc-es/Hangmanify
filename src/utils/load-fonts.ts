import * as Font from 'expo-font';

import { APP_FONTS } from 'assets/fonts';

import { FONTS } from 'src/constants/styles';

export const LoadFonts = async (): Promise<void> => {
  try {
    await Font.loadAsync({
      [FONTS.Nunito.Bold]: APP_FONTS.nunito.bold,
      [FONTS.Nunito.Light]: APP_FONTS.nunito.light,
      [FONTS.Nunito.Medium]: APP_FONTS.nunito.medium,
      [FONTS.Nunito.Regular]: APP_FONTS.nunito.regular,
      [FONTS.Nunito.SemiBold]: APP_FONTS.nunito.semibold,
      [FONTS.Poppins.Bold]: APP_FONTS.poppins.bold,
      [FONTS.Poppins.Light]: APP_FONTS.poppins.light,
      [FONTS.Poppins.Medium]: APP_FONTS.poppins.medium,
      [FONTS.Poppins.Regular]: APP_FONTS.poppins.regular,
      [FONTS.Poppins.SemiBold]: APP_FONTS.poppins.semibold,
    });
  } catch (error) {
    console.error('Error loading fonts', error);
  }
};
