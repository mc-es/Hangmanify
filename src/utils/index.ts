import { cn } from './cn';
import { dimensions } from './dimensions';
import { helpers } from './helpers';

export { cn, dimensions, helpers };
