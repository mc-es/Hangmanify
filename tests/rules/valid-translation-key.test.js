/**
 * @fileoverview Tests for valid-translation-key rule.
 * @see https://github.com/mces58/Hangmanify/blob/master/guides/rules/valid-translation-key.md
 */
import { vol } from 'memfs';

import rule from 'plugins/rules/valid-translation-key.mjs';

import { runTest } from './utils.js';

// Mock the built-in 'fs' module with 'memfs' to simulate a virtual in-memory file system.
// This prevents tests from relying on the real file system.
jest.mock('fs', () => require('memfs').fs);

// Before each test, populate the virtual file system with a fake translation-keys.d.ts file.
// This simulates the TranslationKeys type structure that the ESLint rule depends on.
beforeEach(() => {
  vol.fromJSON({
    'src/constants/localization/translation-keys.d.ts': `
      export type TranslationKeys = {
        greetings: {
          hi: string;
          welcome: string;
        };
        login: {
          username: string;
          password: string;
        };
        profile: {
          avatar: {
            upload: string;
            delete: string;
          };
        };
      };
    `,
  });
});

afterEach(() => {
  // After each test, reset the virtual file system to ensure isolation between tests.
  // This prevents state from leaking across tests.
  vol.reset();
});

const validCases = [
  {
    name: 'Valid key - greetings.hi',
    filename: 'src/screens/Greetings.tsx',
    code: `t('greetings.hi')`,
  },
  {
    name: 'Valid key - nested avatar.upload',
    filename: 'src/screens/Profile.tsx',
    code: `t('profile.avatar.upload')`,
  },
  {
    name: 'Valid key - login.username',
    filename: 'src/screens/Login.tsx',
    code: `t('login.username')`,
  },
  {
    name: 'Valid deeply nested key - profile.avatar.delete',
    filename: 'src/screens/Profile.tsx',
    code: `t('profile.avatar.delete')`,
  },
  {
    name: 'Valid key with single level - greetings.welcome',
    filename: 'src/screens/Greetings.tsx',
    code: `t('greetings.welcome')`,
  },
  {
    name: 'Valid call with other arguments',
    filename: 'src/screens/Login.tsx',
    code: `t('login.password', { fallback: 'pass' })`,
  },
  {
    name: 'Valid usage in JSX expression',
    filename: 'src/screens/Login.tsx',
    code: `<Text>{t('login.username')}</Text>`,
  },
  {
    name: 'Valid usage - variable argument',
    filename: 'src/screens/Dynamic.tsx',
    code: `const key = 'login.username'; t(key);`,
  },
  {
    name: 'Valid usage - template literal',
    filename: 'src/screens/Dynamic.tsx',
    code: 't(`login.${field}`);',
  },
];

const invalidCases = [
  {
    name: 'Invalid key - greetings.bye',
    filename: 'src/screens/Greetings.tsx',
    code: `t('greetings.bye')`,
    expectedErrors: 1,
  },
  {
    name: 'Invalid key - profile.avatar.remove (not defined)',
    filename: 'src/screens/Profile.tsx',
    code: `t('profile.avatar.remove')`,
    expectedErrors: 1,
  },
  {
    name: 'Invalid key - completely unknown section',
    filename: 'src/screens/Settings.tsx',
    code: `t('settings.theme')`,
    expectedErrors: 1,
  },
  {
    name: 'Invalid key - profile.avatar',
    filename: 'src/screens/Profile.tsx',
    code: `t('profile.avatar')`,
    expectedErrors: 1,
  },
  {
    name: 'Invalid nested key - profile.avatar.upload.icon',
    filename: 'src/screens/Profile.tsx',
    code: `t('profile.avatar.upload.icon')`,
    expectedErrors: 1,
  },
  {
    name: 'Invalid partial key - login',
    filename: 'src/screens/Login.tsx',
    code: `t('login')`,
    expectedErrors: 1,
  },
];

runTest('valid-translation-key', rule, validCases, invalidCases);
