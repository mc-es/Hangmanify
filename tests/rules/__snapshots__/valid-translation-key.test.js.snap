// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`valid-translation-key rule invalid cases Invalid key - completely unknown section 1`] = `
[
  {
    "column": 3,
    "endColumn": 19,
    "endLine": 1,
    "line": 1,
    "message": "The translation key "settings.theme" is not defined in TranslationKeys.Did you mean: "greetings", "login", "profile"?",
    "messageId": "invalidKey",
    "nodeType": "Literal",
    "ruleId": "@custom-typescript-eslint/valid-translation-key",
    "severity": 2,
  },
]
`;

exports[`valid-translation-key rule invalid cases Invalid key - greetings.bye 1`] = `
[
  {
    "column": 3,
    "endColumn": 18,
    "endLine": 1,
    "line": 1,
    "message": "The translation key "greetings.bye" is not defined in TranslationKeys.Did you mean: "greetings.hi", "greetings.welcome"?",
    "messageId": "invalidKey",
    "nodeType": "Literal",
    "ruleId": "@custom-typescript-eslint/valid-translation-key",
    "severity": 2,
  },
]
`;

exports[`valid-translation-key rule invalid cases Invalid key - profile.avatar 1`] = `
[
  {
    "column": 3,
    "endColumn": 19,
    "endLine": 1,
    "line": 1,
    "message": "The translation key "profile.avatar" is not defined in TranslationKeys.",
    "messageId": "invalidKey",
    "nodeType": "Literal",
    "ruleId": "@custom-typescript-eslint/valid-translation-key",
    "severity": 2,
  },
]
`;

exports[`valid-translation-key rule invalid cases Invalid key - profile.avatar.remove (not defined) 1`] = `
[
  {
    "column": 3,
    "endColumn": 26,
    "endLine": 1,
    "line": 1,
    "message": "The translation key "profile.avatar.remove" is not defined in TranslationKeys.Did you mean: "profile.avatar.delete", "profile.avatar.upload"?",
    "messageId": "invalidKey",
    "nodeType": "Literal",
    "ruleId": "@custom-typescript-eslint/valid-translation-key",
    "severity": 2,
  },
]
`;

exports[`valid-translation-key rule invalid cases Invalid nested key - profile.avatar.upload.icon 1`] = `
[
  {
    "column": 3,
    "endColumn": 31,
    "endLine": 1,
    "line": 1,
    "message": "The translation key "profile.avatar.upload.icon" is not defined in TranslationKeys.",
    "messageId": "invalidKey",
    "nodeType": "Literal",
    "ruleId": "@custom-typescript-eslint/valid-translation-key",
    "severity": 2,
  },
]
`;

exports[`valid-translation-key rule invalid cases Invalid partial key - login 1`] = `
[
  {
    "column": 3,
    "endColumn": 10,
    "endLine": 1,
    "line": 1,
    "message": "The translation key "login" is not defined in TranslationKeys.",
    "messageId": "invalidKey",
    "nodeType": "Literal",
    "ruleId": "@custom-typescript-eslint/valid-translation-key",
    "severity": 2,
  },
]
`;
